nautobot/dcim/models/devices.py

```diff
@@ -818,21 +818,20 @@ def clean(self):
                 }
             )
 
-        # Validate device software version has a software image file that matches the device's device type or is a default image
-        if self.software_version is not None and not any(
-            (
-                self.software_version.software_image_files.filter(device_types=self.device_type).exists(),
-                self.software_version.software_image_files.filter(default_image=True).exists(),
-            )
-        ):
-            raise ValidationError(
-                {
-                    "software_version": (
-                        f"No software image files for version '{self.software_version}' are "
-                        f"valid for device type {self.device_type}."
+        # If any software image file is specified, validate that
+        # each of the software image files belongs to the device's device type or is a default image
+
+        if self.software_image_files.exists():
```

Comment:
It occurs to me that it'd be "cleaner" if we had a proper M2M model for device-to-software-image-file, as really this logic belongs on the `clean()` for *that* model, rather than on Device itself, so that something like `device.software_image_files.set([invalid_file])` would itself raise the validation error independent of needing to call `device.validated_save()`. This is probably a good step in the right direction as-is though.

Reply:
https://github.com/nautobot/nautobot/issues/6110